pkg/controlplane/controller/defaultservicecidr/default_servicecidr_controller.go

```diff
@@ -0,0 +1,219 @@
+
+	interval time.Duration
+}
+
+// Start will not return until the default ServiceCIDR exists or stopCh is closed.
+func (c *Controller) Start(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	c.eventBroadcaster.StartStructuredLogging(0)
+	c.eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: c.client.CoreV1().Events("")})
+	defer c.eventBroadcaster.Shutdown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	go c.serviceCIDRInformer.Run(stopCh)
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.serviceCIDRsSynced) {
+		return
+	}
+
+	go wait.Until(c.sync, c.interval, stopCh)
+
+	select {
+	case <-stopCh:
+	case <-c.readyCh:
+	}
+}
+
+func (c *Controller) sync() {
+	// check if the default ServiceCIDR already exist
+	serviceCIDR, err := c.serviceCIDRLister.Get(DefaultServiceCIDRName)
+	// if exists
+	if err == nil {
+		c.setReady()
+		c.syncStatus(serviceCIDR)
+		return
+	}
+
+	// unknown error
+	if !apierrors.IsNotFound(err) {
+		klog.Infof("error trying to obtain the default ServiceCIDR: %v", err)
+		return
+	}
+
+	// default ServiceCIDR does not exist
+	klog.Infof("Creating default ServiceCIDR, ipv4: %q ipv6: %q", c.ipv4CIDR, c.ipv6CIDR)
+	serviceCIDR = &networkingapiv1alpha1.ServiceCIDR{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: DefaultServiceCIDRName,
+		},
+		Spec: networkingapiv1alpha1.ServiceCIDRSpec{
+			IPv4: c.ipv4CIDR,
+			IPv6: c.ipv6CIDR,
+		},
+	}
+	serviceCIDR, err = c.client.NetworkingV1alpha1().ServiceCIDRs().Create(context.Background(), serviceCIDR, metav1.CreateOptions{})
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		klog.Infof("error creating default ServiceCIDR: %v", err)
+		c.eventRecorder.Eventf(serviceCIDR, v1.EventTypeWarning, "KubernetesDefaultServiceCIDRError", "The default ServiceCIDR can not be created")
+		return
+	}
+
+	c.setReady()
+	c.syncStatus(serviceCIDR)
+}
+
+func (c *Controller) setReady() {
+	select {
+	case <-c.readyCh:
+	default:
+		close(c.readyCh)
+	}
+}
+
+func (c *Controller) syncStatus(serviceCIDR *networkingapiv1alpha1.ServiceCIDR) {
+	// don't sync the status of the ServiceCIDR if is being deleted,
+	// deletion must be handled by the controller-manager
+	if !serviceCIDR.GetDeletionTimestamp().IsZero() {
+		return
+	}
+
+	// This controller will set the Ready condition to true if the Ready condition
+	// does not exist and the CIDR values match this controller CIDR values.
+	for _, condition := range serviceCIDR.Status.Conditions {
+		if condition.Type == networkingapiv1alpha1.ServiceCIDRConditionReady {
+			if condition.Status == metav1.ConditionTrue {
+				return
+			}
+			klog.Infof("default ServiceCIDR condition Ready is not True: %v", condition.Status)
+			c.eventRecorder.Eventf(serviceCIDR, v1.EventTypeWarning, condition.Reason, condition.Message)
+			return
+		}
+	}
+	// set status to ready if the ServiceCIDR matches this configuration
+	if c.ipv4CIDR == serviceCIDR.Spec.IPv4 &&
```

Comment:
In case of multiple API servers, if just one of them has the wrong config, won't this condition flip back and forth, depending on who wrote it last?

Reply:
the condition is only set to true here, in theory (Dan already pointed it out) this controller should not event have to set the condition, but I wanted to be paranoid and cover the case that the controller manager is not running and not setting the condition.

It can not flip because it only goes to true if it matches the conditions, apiservers with different configfs just will exist